mmengine/runner/deepspeed_runner.py

```diff
@@ -0,0 +1,771 @@
+                epochs. Defaults to True.
+            backend_args (dict, optional): Arguments to instantiate the
+                preifx of uri corresponding backend. Defaults to None.
+                New in v0.2.0.
+        """
+        if meta is None:
+            meta = {}
+        elif not isinstance(meta, dict):
+            raise TypeError(
+                f'meta should be a dict or None, but got {type(meta)}')
+
+        if by_epoch:
+            # self.epoch increments 1 after
+            # `self.call_hook('after_train_epoch)` but `save_checkpoint` is
+            # called by `after_train_epoch`` method of `CheckpointHook` so
+            # `epoch` should be `self.epoch + 1`
+            meta.update(epoch=self.epoch + 1, iter=self.iter)
+        else:
+            meta.update(epoch=self.epoch, iter=self.iter + 1)
+
+        file_client = FileClient.infer_client(file_client_args, out_dir)
+        filepath = file_client.join_path(out_dir, filename)
+
+        meta.update(
+            cfg=self.cfg.pretty_text,
+            seed=self.seed,
+            experiment_name=self.experiment_name,
+            time=time.strftime('%Y%m%d_%H%M%S', time.localtime()),
+            mmengine_version=mmengine.__version__ + get_git_hash(),
+            deepspeed_config=self.ds_config,
+            deepspeed_version=deepspeed.git_version_info.version)
+
+        if hasattr(self.train_dataloader.dataset, 'metainfo'):
+            meta.update(dataset_meta=self.train_dataloader.dataset.metainfo)
+
+        # TODO: Support DeepSpeed-MoE
+        if self.model.has_moe_layers:
+            raise NotImplementedError('DeepSpeed-MoE is not supported yet.')
+
+        if is_model_wrapper(self.model):
+            model = self.model.module
+        else:
+            model = self.model
+
+        if self.model.zero_optimization_partition_weights():
+            # Prepare for checkpoint save by
+            # ensuring all parameters are partitioned
+            self.model.optimizer.checkpoint_event_prologue()
+
+        checkpoint = {
+            'meta': meta,
+            'message_hub': self.message_hub.state_dict(),
+        }
+        # save optimizer state dict to checkpoint
+        if save_optimizer:
+            if not self.model.zero_optimization():
+                checkpoint['optimizer'] = self.optim_wrapper.state_dict()
+            else:
+                self.consolidate_state_dict(self.optim_wrapper.state_dict())
+                # Only the main process needs to load the optimizer's state.
+                optim_state = self.get_zero_state_dict()
+                checkpoint['optimizer'] = optim_state
+
+        # model state is stored after pulling optimizer state to handle ZeRO 3.
+        checkpoint['state_dict'] = weights_to_cpu(self.get_state_dict(model))
+
+        if self.model.zero_optimization_partition_weights():
+            self.model.optimizer.checkpoint_event_epilogue()
```

Comment:
This logic is from https://github.com/microsoft/DeepSpeed/blob/211055216792cbb52ab6d355f698c194f9c55efb/deepspeed/runtime/engine.py#L2885.